openpype/hosts/maya/plugins/publish/validate_resolution.py

```diff
@@ -0,0 +1,81 @@
+
+    order = pyblish.api.ValidatorOrder - 0.01
+    families = ["renderlayer"]
+    hosts = ["maya"]
+    label = "Validate Resolution"
+    actions = [RepairAction]
+    optional = True
+
+    def process(self, instance):
+        if not self.is_active(instance.data):
+            return
+        width, height, pixelAspect = self.get_db_resolution(instance)
+        current_renderer = cmds.getAttr(
+            "defaultRenderGlobals.currentRenderer")
+        layer = instance.data["renderlayer"]
+        if current_renderer == "vray":
+            vray_node = "vraySettings"
+            if cmds.objExists(vray_node):
+                control_node = vray_node
+                current_width = lib.get_attr_in_layer(
+                    "{}.width".format(control_node), layer=layer)
+                current_height = lib.get_attr_in_layer(
+                    "{}.height".format(control_node), layer=layer)
+                current_pixelAspect = lib.get_attr_in_layer(
+                    "{}.pixelAspect".format(control_node), layer=layer
+                )
+            else:
+                raise PublishValidationError(
+                    "Can't set VRay resolution because there is no node "
+                    "named: `%s`" % vray_node)
+        else:
+            current_width = lib.get_attr_in_layer(
+                "defaultResolution.width", layer=layer)
+            current_height = lib.get_attr_in_layer(
+                "defaultResolution.height", layer=layer)
+            current_pixelAspect = lib.get_attr_in_layer(
+                "defaultResolution.pixelAspect", layer=layer
+            )
+        if current_width != width or current_height != height:
+            raise PublishValidationError(
+                "Render resolution {}x{} does not match asset resolution {}x{}".format(         # noqa:E501
+                    current_width, current_height, width, height
+                ))
+        if current_pixelAspect != pixelAspect:
+            raise PublishValidationError(
+                "Render pixel aspect {} does not match asset pixel aspect {}".format(            # noqa:E501
+                    current_pixelAspect, pixelAspect
+                ))
+
+    def get_db_resolution(self, instance):
+        asset_doc = instance.data["assetEntity"]
+        project_doc = instance.context.data["projectEntity"]
+        for data in [asset_doc["data"], project_doc["data"]]:
+            if "resolutionWidth" in data and (
+                "resolutionHeight" in data and "pixelAspect" in data
+            ):
+                width = data["resolutionWidth"]
+                height = data["resolutionHeight"]
+                pixelAspect = data["pixelAspect"]
+                return int(width), int(height), int(pixelAspect)
+
+        # Defaults if not found in asset document or project document
+        return 1920, 1080, 1
+
+    @classmethod
+    def repair(cls, instance):
+        return reset_scene_resolution()
```

Comment:
Issue still exists that this does not repair the width/height if the renderlayer were using a renderlayer override on resolution.

However, we _can_ create an issue for that so we can track this 'bug' for later and get this first pass of the feature merged in for community testing.

Reply:
maybe we can build some new function fix it by referencing` lib.get_attr_in_layer`

Reply:
Well, the logic to reset scene resolution can grow complicated across multiple renderers since each has their own settings, etc. - That's why we have abstracted that into the `lib_rendersettings` so there's a hopefully somewhat ok API to reset all different renders, etc.

Anyway, I believe however that `reset_scene_resolution` just sets the current value. If you're **inside a renderlayer with a renderlayer override** I believe setting the value equals to updating the layer override value.

We could potentially use [this function `get_attr_overrides`](https://github.com/ynput/OpenPype/blob/a93b63561814ed2d6bed0b97503b3f51acd2c757/openpype/hosts/maya/api/lib_rendersetup.py#L186) to detect if a layer has overrides and if so switch to the renderlayer.

**pseudocode**

```python
attr = "attr"
current_layer = "x"
layer = "y"

# We are in the renderlayer - we can just set the value directly
# even if it has a layer override
if current_layer == layer:
    # Just reset since we're in the layer
    reset_scene_resolution()
    return

# If the layer has an override or the current layer has override
# then we must switch to the layer to ensure we change the
# correct value that affects that layer
if (
    get_attr_overrides(attr, layer=layer) or
    get_attr_overrides(attr, layer=current_layer)
):
    with lib.renderlayer(layer):
        reset_scene_resolution()
        return

# We are not in the same layer but both don't have layer overrides
# so whatever we set in our current layer will affect that layer too
reset_scene_resolution()
```